cirq/ops/controlled_operation.py

```diff
@@ -0,0 +1,106 @@
+# limitations under the License.
+from typing import Union, Any, Optional
+
+import numpy as np
+
+from cirq import protocols, linalg, value
+from cirq.ops import raw_types
+from cirq.type_workarounds import NotImplementedType
+
+
+@value.value_equality
+class ControlledOperation(raw_types.Operation):
+    def __init__(self,
+                 control: raw_types.QubitId,
+                 sub_operation: raw_types.Operation):
+        self.control = control
+        self.sub_operation = sub_operation
+
+    @property
+    def qubits(self):
+        return (self.control,) + self.sub_operation.qubits
+
+    def with_qubits(self, *new_qubits):
+        return ControlledOperation(
+            new_qubits[0],
+            self.sub_operation.with_qubits(*new_qubits[1:]))
+
+    def _decompose_(self):
+        result = protocols.decompose_once(self.sub_operation, NotImplemented)
+        if result is NotImplemented:
+            return NotImplemented
+
+        return [ControlledOperation(self.control, op) for op in result]
+
+    def _value_equality_values_(self):
+        return tuple([self.control, self.sub_operation])
+
+    def _has_unitary_(self) -> bool:
+        return protocols.has_unitary(self.sub_operation)
+
+    def _unitary_(self) -> Union[np.ndarray, NotImplementedType]:
+        sub_matrix = protocols.unitary(self.sub_operation, None)
+        if sub_matrix is None:
+            return NotImplemented
+        return linalg.block_diag(np.eye(sub_matrix.shape[0]), sub_matrix)
+
+    def __str__(self):
+        return 'C({})'.format(self.control) + str(self.sub_operation)
+
+    def __repr__(self):
+        return 'cirq.ControlledOperation(control={!r},' \
+               ' sub_operation={!r})'.format(self.control, self.sub_operation)
+
+    def _is_parameterized_(self) -> bool:
+        return protocols.is_parameterized(self.sub_operation)
+
+    def _resolve_parameters_(self, resolver):
+        new_sub_op = protocols.resolve_parameters(self.sub_operation, resolver)
+        return ControlledOperation(self.control, new_sub_op)
+
+    def _trace_distance_bound_(self) -> float:
+        return protocols.trace_distance_bound(self.sub_operation)
+
+    def __pow__(self, exponent: Any) -> 'ControlledOperation':
+        new_sub_op = protocols.pow(self.sub_operation,
+                                   exponent,
+                                   NotImplemented)
+        if new_sub_op is NotImplemented:
+            return NotImplemented
+        return ControlledOperation(self.control, new_sub_op)
+
+    def _circuit_diagram_info_(self,
+                               args: protocols.CircuitDiagramInfoArgs
+                               ) -> Optional[protocols.CircuitDiagramInfo]:
+
+        sub_args = protocols.CircuitDiagramInfoArgs(
+            known_qubit_count=(args.known_qubit_count - 1
+                               if args.known_qubit_count is not None else None),
+            known_qubits=(tuple(b for b in args.known_qubits
```

Comment:
The known qubits have to be in order, so `args.known_qubits[1:]` will work when it's not none.

Reply:
ack